File: geo_tuner/core/safety.py
"""Independent safety monitor for in-flight tuning episodes.

Pure logic (no ROS) so it is unit-testable. The conductor node feeds it
odometry samples; any violation aborts the tuning session into a hover
hold with the last known-safe gains.
"""

from __future__ import annotations

import math
from collections import deque
from dataclasses import dataclass, field
from enum import Enum


class Violation(Enum):
    TILT = "tilt limit exceeded"
    POS_ERROR = "position error bound exceeded"
    VELOCITY = "velocity bound exceeded"
    ALTITUDE_LOW = "below minimum altitude"
    ALTITUDE_HIGH = "above maximum altitude"
    OSCILLATION = "oscillation detected (body rate energy)"
    ODOM_STALE = "odometry stale"


@dataclass
class SafetyLimits:
    max_tilt: float = 0.6          # rad (~34 deg), > controller max_tilt_angle
    max_pos_error: float = 2.0     # m, distance from active setpoint
    max_velocity: float = 4.0      # m/s
    min_altitude: float = 1.0      # m
    max_altitude: float = 40.0     # m
    odom_timeout: float = 0.3      # s
    # Oscillation detector: RMS of mean-removed body rates over the window
    osc_window: float = 2.0        # s
    osc_rate_rms: float = 1.2      # rad/s, roll+pitch combined
    osc_yaw_rate_rms: float = 1.5  # rad/s, yaw alone (drag-torque axis)
    osc_min_samples: int = 20


@dataclass
class OdomSample:
    t: float                       # s
    pos: tuple                     # (x, y, z) m
    vel: tuple                     # (vx, vy, vz) m/s
    quat: tuple                    # (w, x, y, z)
    body_rates: tuple = (0.0, 0.0, 0.0)  # rad/s


def tilt_from_quat(q: tuple) -> float:
    """Angle between body z and world z, from (w,x,y,z) quaternion."""
    w, x, y, z = q
    # R[2][2] of the rotation matrix
    r22 = 1.0 - 2.0 * (x * x + y * y)
    return math.acos(max(-1.0, min(1.0, r22)))


@dataclass
class SafetyMonitor:
    limits: SafetyLimits = field(default_factory=SafetyLimits)
    _rates: deque = field(default_factory=deque)   # (t, wx, wy, wz)
    _last_t: float | None = None

    def check(self, s: OdomSample, setpoint: tuple | None) -> list[Violation]:
        """Feed one odometry sample; returns violations (empty = safe)."""
        v: list[Violation] = []
        self._last_t = s.t

        if tilt_from_quat(s.quat) > self.limits.max_tilt:
            v.append(Violation.TILT)

        speed = math.sqrt(sum(c * c for c in s.vel))
        if speed > self.limits.max_velocity:
            v.append(Violation.VELOCITY)

        if s.pos[2] < self.limits.min_altitude:
            v.append(Violation.ALTITUDE_LOW)
        if s.pos[2] > self.limits.max_altitude:
            v.append(Violation.ALTITUDE_HIGH)

        if setpoint is not None:
            err = math.sqrt(sum((a - b) ** 2 for a, b in zip(s.pos, setpoint)))
            if err > self.limits.max_pos_error:
                v.append(Violation.POS_ERROR)

        # Rolling body-rate oscillation energy
        self._rates.append((s.t, *s.body_rates))
        t0 = s.t - self.limits.osc_window
        while self._rates and self._rates[0][0] < t0:
            self._rates.popleft()
        if len(self._rates) >= self.limits.osc_min_samples:
            n = len(self._rates)

            def _var(axis):
                vals = [r[axis] for r in self._rates]
                mean = sum(vals) / n
                return sum((x - mean) ** 2 for x in vals) / n

            if math.sqrt(_var(1) + _var(2)) > self.limits.osc_rate_rms:
                v.append(Violation.OSCILLATION)
            if math.sqrt(_var(3)) > self.limits.osc_yaw_rate_rms:
                v.append(Violation.OSCILLATION)

        return v

    def check_stale(self, now: float) -> list[Violation]:
        if self._last_t is None:
            return []
        if now - self._last_t > self.limits.odom_timeout:
            return [Violation.ODOM_STALE]
        return []

    def reset(self):
        self._rates.clear()
        self._last_t = None

File: src/core/safety.cpp
#include "geo_tuner/core/safety.hpp"

#include <algorithm>
#include <cmath>

namespace geo_tuner
{

const char * to_string(Violation v)
{
  switch (v) {
    case Violation::TILT: return "tilt limit exceeded";
    case Violation::POS_ERROR: return "position error bound exceeded";
    case Violation::VELOCITY: return "velocity bound exceeded";
    case Violation::ALTITUDE_LOW: return "below minimum altitude";
    case Violation::ALTITUDE_HIGH: return "above maximum altitude";
    case Violation::OSCILLATION: return "oscillation detected (body rate energy)";
    case Violation::ODOM_STALE: return "odometry stale";
  }
  return "unknown violation";
}

double tilt_from_quat(const std::array<double, 4> & q)
{
  const double x = q[1], y = q[2];
  // R[2][2] of the rotation matrix
  const double r22 = 1.0 - 2.0 * (x * x + y * y);
  return std::acos(std::clamp(r22, -1.0, 1.0));
}

std::vector<Violation> SafetyMonitor::check(
  const OdomSample & s, const std::optional<std::array<double, 3>> & setpoint,
  bool check_min_altitude)
{
  std::vector<Violation> v;
  last_t_ = s.t;

  if (tilt_from_quat(s.quat) > limits_.max_tilt) {
    v.push_back(Violation::TILT);
  }

  const double speed = std::sqrt(
    s.vel[0] * s.vel[0] + s.vel[1] * s.vel[1] + s.vel[2] * s.vel[2]);
  if (speed > limits_.max_velocity) {
    v.push_back(Violation::VELOCITY);
  }

  const double alt = altitude_of(s);
  if (check_min_altitude && alt < limits_.min_altitude) {
    v.push_back(Violation::ALTITUDE_LOW);
  }
  if (alt > limits_.max_altitude) {
    v.push_back(Violation::ALTITUDE_HIGH);
  }

  if (setpoint) {
    double sq = 0.0;
    for (int i = 0; i < 3; ++i) {
      const double d = s.pos[i] - (*setpoint)[i];
      sq += d * d;
    }
    if (std::sqrt(sq) > limits_.max_pos_error) {
      v.push_back(Violation::POS_ERROR);
    }
  }

  // Rolling body-rate oscillation energy
  rates_.push_back({s.t, s.body_rates[0], s.body_rates[1], s.body_rates[2]});
  const double t0 = s.t - limits_.osc_window;
  while (!rates_.empty() && rates_.front()[0] < t0) {
    rates_.pop_front();
  }
  if (static_cast<int>(rates_.size()) >= limits_.osc_min_samples) {
    const double n = static_cast<double>(rates_.size());
    auto var = [this, n](int axis) {
        double mean = 0.0;
        for (const auto & r : rates_) {mean += r[axis];}
        mean /= n;
        double acc = 0.0;
        for (const auto & r : rates_) {
          const double d = r[axis] - mean;
          acc += d * d;
        }
        return acc / n;
      };
    if (std::sqrt(var(1) + var(2)) > limits_.osc_rate_rms) {
      v.push_back(Violation::OSCILLATION);
    }
    if (std::sqrt(var(3)) > limits_.osc_yaw_rate_rms) {
      v.push_back(Violation::OSCILLATION);
    }
  }

  return v;
}

std::vector<Violation> SafetyMonitor::check_stale(double now) const
{
  if (!last_t_) {return {};}
  if (now - *last_t_ > limits_.odom_timeout) {
    return {Violation::ODOM_STALE};
  }
  return {};
}

void SafetyMonitor::reset()
{
  rates_.clear();
  last_t_.reset();
}

}  // namespace geo_tuner
